### PR TITLE
Keras metric callback

### PR DESCRIPTION
**Pull request by Rocketknight1**
Hey, here's the unfinished Keras metric PR, co-authored with @merveenoyan!

Still to do:

1) The outputs from predict/generate are not being concatenated properly, this will be quick to fix! (Edit: Done)
2) I haven't tested prediction with generate at all. `generate()` is still very slow, but that's a separate PR :fearful: 
3) More broad testing with multiple metric functions.

### CODE REVIEW

**Comment by merveenoyan** [comment]
I feel like it makes sense to test on three notebooks for best coverage.
- Token clf
- Translation
- Masked language modeling
If you approve I’ll test for these ones. 
@Rocketknight1

File: src/transformers/keras_callbacks.py [hunk 1]
```diff
@@ -16,6 +18,132 @@
...
+        val_dataset: Validation data to be used to evaluate the model at
+        the end of the epoch.
+        batch_size: Batch size.
+        labels: Labels.
+    """
+
+    def __init__(
+        self,
+        metric_fn: Callable,
+        val_dataset: Union[tf.data.Dataset, np.ndarray, tf.Tensor, tuple, dict],
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:36]: The trainer uses `eval_dataset` but maybe Keras uses `val_dataset` more consistently?

File: src/transformers/keras_callbacks.py [hunk 2]
```diff
@@ -16,6 +18,132 @@
...
+        label_list = []
+
+        # The whole predict/generate loop is handled inside this method
+        for batch in self.val_dataset:
+            if isinstance(batch, tuple):
+                batch, labels = batch
+            else:
+                labels = None
+            if self.predict_with_generate:
+                predictions = self.model.generate(batch)
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:113]: I think you might be passing dangerous keys here. On the trainer side, we only pass along the inputs that are in the `tokenizer.input_names`. As we saw before, passing `decoder_input_ids` for instance hurts performance, and they might be in the processed dataset.

Also, we allow generation kwargs in the trainer (`max_length`, `num_beans` for instance) that could be useful for the user to control here.

**Review by sgugger — APPROVED** [review]
Thanks a lot for the PR! I think there are a few points to polish but it should be ready to merge soon!

File: src/transformers/keras_callbacks.py [hunk 3]
```diff
@@ -16,6 +18,132 @@
...
+class KerasMetricCallback(Callback):
+    """
+    Callback to prompt metrics at the end of every epoch.
+
+    Args:
+        metric_fn: Metric function provided by the user.
+        val_dataset: Validation data to be used to evaluate the model at
+        the end of the epoch.
+        batch_size: Batch size.
+        labels: Labels.
```
[the remark this answers is not in the document]
Comment (merveenoyan) [reply src/transformers/keras_callbacks.py:30]: On it, thanks!

**Review by LysandreJik — APPROVED** [review]
Very nice! You mentioned that this will only work with the latest version of TensorFlow, should there be a check to ensure that users trying to use this for earlier versions are not surprised when it doesn't work for their setup?

**Comment by Rocketknight1** [comment]
@LysandreJik Absolutely yes, that's getting added!

File: src/transformers/keras_callbacks.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/keras_callbacks.py:36]: I'm fine to change this to `eval_dataset` for internal consistency!

File: src/transformers/keras_callbacks.py (hunk 2 above)
Comment (Rocketknight1) [reply src/transformers/keras_callbacks.py:113]: Either way, done!

File: src/transformers/keras_callbacks.py [hunk 4]
```diff
@@ -16,6 +18,136 @@
...
+        label_list = []
+
+        # The whole predict/generate loop is handled inside this method
+        for batch in self.eval_dataset:
+            if isinstance(batch, tuple):
+                batch, labels = batch
+            else:
+                labels = None
+            if isinstance(batch, dict):
+                batch = {key: array for key, array in batch.items() if key in self.model_input_names}
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:115]: This should only be in the `predict_with_generate` block ;-)

**Review by sgugger — COMMENTED** [review]
You'll need to rebase on master and convert all docstrings to Markdown :-)

File: src/transformers/keras_callbacks.py (hunk 4 above)
Comment (sgugger) [reply src/transformers/keras_callbacks.py:115]: For training the keys are not harmful (yes it's a bit of a hack, I know).

In general we should trust the user provided the proper keys and not forcibly remove any he passed. `predict_with_generate` has some dark magic that I hope we can remove one day, but in the meantime, let's only remove keys there :-)

Comment (Rocketknight1) [reply src/transformers/keras_callbacks.py:115]: Sure!

File: src/transformers/keras_callbacks.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/keras_callbacks.py]: Done!

File: src/transformers/keras_callbacks.py [hunk 5]
```diff
@@ -16,6 +18,185 @@
...
+            These contain the model's outputs and matching labels from the dataset. It should return a dict mapping
+            metric names to numerical values.
+        eval_dataset (`tf.data.Dataset` or `dict` or `tuple` or `np.ndarray` or `tf.Tensor`):
+            Validation data to be used to generate predictions for the `metric_fn`.
+        metric_fn_kwargs (`dict`, *optional*):
+            Additional keyword arguments to be passed to the metric_fn.
+        tokenizer ([`PretrainedTokenizerBase`], *optional*):
+            Tokenizer used to validate column names to be passed to the generate() function. Required only if
+            predict_with_generate is True.
+        output_cols: (`List[str], *optional*):
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:62]:
```suggestion
        output_cols (`List[str], *optional*):
```

File: src/transformers/keras_callbacks.py [hunk 6]
```diff
@@ -16,6 +18,185 @@
...
+            A list of columns to be retained from the model output as the predictions. Defaults to all.
+        label_cols: ('`List[str]`, *optional*'):
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:64]:
```suggestion
        label_cols ('`List[str]`, *optional*'):
```

File: src/transformers/keras_callbacks.py [hunk 7]
```diff
@@ -16,6 +18,185 @@
...
+            A list of columns to be retained from the input dataset as the labels. Will be autodetected if this is not
+            supplied.
+        batch_size (`int`, *optional*):
+            Batch size. Only used when the data is not a pre-batched `tf.data.Dataset`.
+        predict_with_generate: (`bool`, *optional*, defaults to *False*):
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:69]:
```suggestion
        predict_with_generate: (`bool`, *optional*, defaults to `False`):
```

File: src/transformers/keras_callbacks.py [hunk 8]
```diff
@@ -16,6 +18,185 @@
...
+        # and may not be directly applicable to any particular use-case
+        decoded_predictions = tokenizer.batch_decode(predictions, skip_special_tokens=True))
+        decoded_labels = tokenizer.batch_decode(labels, skip_special_tokens=True)
+        result = rouge_metric.compute(predictions=decoded_predictions, references=decoded_labels)
+        return {key: value.mid.fmeasure * 100 for key, value in result.items()}
+    ```
+
+    The above function will return a dict containing values which will be logged like any other Keras metric:
+
+    ```py
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:46]:
```suggestion
    ```
```
It's an output so would leave without the py.

File: src/transformers/keras_callbacks.py [hunk 9]
```diff
@@ -16,6 +18,185 @@
...
+        output_cols: Optional[List[str]] = None,
+        label_cols: Optional[List[str]] = None,
+        batch_size: Optional[int] = None,
+        predict_with_generate: Optional[bool] = False,
+    ):
+        super().__init__()
+        self.metric_fn = metric_fn
+        self.batch_size = batch_size
+        if predict_with_generate and tokenizer is None:
+            raise ValueError("A tokenizer is required when using predict_with_generate!")
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:89]: Not necessarily. In the `Trainer`, we just take `"input_ids"` if no tokenizer is passed.

**Review by sgugger — APPROVED** [review]
A few last nits. And I think it's still missing the check for the minimum version of TensorFlow?

File: src/transformers/keras_callbacks.py (hunk 9 above)
Comment (Rocketknight1) [reply src/transformers/keras_callbacks.py]: Fixed!
